lib/handler/pool/state.ex

```diff
@@ -54,59 +62,52 @@ defmodule Handler.Pool.State do
           {:ok, t(), reference()} | {:reject, exception()}
   def start_worker(state, fun, opts, from_pid) do
     bytes_requested = max_heap_bytes(opts)
-    name = task_name(opts)
 
     with :ok <- check_committed_resources(state, bytes_requested),
-         {:ok, ref, task_pid} <- kickoff_new_task(state, fun, opts) do
-      new_state = commit_resources(state, ref, bytes_requested, from_pid, task_pid, name)
-
+         {:ok, ref, worker} <- kickoff_new_task(state, fun, opts, from_pid) do
+      new_state = commit_resources(state, ref, worker)
       {:ok, new_state, ref}
     end
   end
 
-  def kill_worker(%State{pool: %Pool{delegate_to: pool}} = state, task_name)
-      when not is_nil(pool) do
-    if has_worker_named(state, task_name) do
-      {:ok, number_killed} = Pool.kill(pool, task_name)
-      {:ok, number_killed, state}
-    else
-      {:ok, 0, state}
-    end
+  @spec kill_worker(t(), String.t()) :: {:ok, t(), non_neg_integer()}
+  def kill_worker(state, task_name) do
+    Enum.reduce(state.workers, {:ok, state, 0}, fn
+      {_ref, %{task_pid: task_pid, task_name: ^task_name}}, {:ok, state, number_killed} ->
+        :ok = Process.send(task_pid, {Pool, :user_killed}, [])
```

Comment:
So the pool will now rely on the handler to commit seppuku, instead of the pool killing the wrapper task directly. When we fixed this yesterday, we used Task.shutdown(w/ brutal_kill)  from the pool instead, and removed all mentions of user killing from Handler.run. I think the end result is the same, do you think using Task.shutdown from the pool would have any drawbacks in comparison to this approach? What I don't like about this is that Handler has to acknowledge that it may belong to a pool and that it's pool may ask it do kill itself.